Add tutorial for Speed Filter (#124)
ros-planning/navigation.ros.org: tutorials/docs/navigation2_with_speed_filter.rst

 - Speed limit expressed in absolute values (e.g. in ``m/s``).

The meaning used by Speed Filter is being read from ``nav2_msgs/CostmapFilterInfo`` messages.
In this tutorial we will use the first type of speed restriction expressed in a percent from maximum robot speed.

.. note::

  For speed restriction expressed in a percent, ``speed_limit`` will be used exactly as a percent belonging to ``[0..100]`` range, not ``[0.0..1.0]`` range.

Create a new image with a PGM/PNG/BMP format: copy ``turtlebot3_world.pgm`` main map which will be used in a world simulation from a navigation2 repository to a new ``speed_mask.pgm`` file. Open ``speed_mask.pgm`` in your favourite raster graphics editor and fill speed restricted areas with grey colors. In our example darker colors will indicate areas with higher speed restriction:

.. image:: images/Navigation2_with_Speed_Filter/drawing_speed_mask.png
    :width: 500px

Area "A" is filled with ``40%`` gray color, area "B" - with ``70%`` gray, that means that speed restriction will take ``100% - 40% = 60%`` in area "A" and ``100% - 70% = 30%`` in area "B" from maximum speed value allowed for this robot.

> We will use ``scale`` map mode with no thresholds. In this mode darker colors will have higher ``OccupancyGrid`` values. E.g. for area "B" with ``70%`` of gray ``OccupancyGrid`` data will be equal to ``70``. So in order to hit the target, we need to choose ``base = 100.0`` and ``multiplier = -1.0``. This will reverse the scale ``OccupancyGrid`` values to a desired one. No thresholds were chosen for the convenience: to have 1:1 full range conversion of lightness value from filter mask -> to speed restriction percent.

Reviewer: ```suggestion
We will use ``scale`` map mode with no thresholds. In this mode darker colors will have higher ``OccupancyGrid`` values. E.g. for area "B" with ``70%`` of gray ``OccupancyGrid`` data will be equal to ``70``. So in order to hit the target, we need to choose ``base = 100.0`` and ``multiplier = -1.0``. This will reverse the scale ``OccupancyGrid`` values to a desired one. No thresholds (``free_thresh`` ``occupied_thresh``) were chosen for the convenience in the ``yaml``  file: to have 1:1 full range conversion of lightness value from filter mask -> to speed restriction percent.
```

.. _navigation2_with_speed_filter:

Navigating with Speed Limits
****************************

- `Overview`_
- `Requirements`_
- `Tutorial Steps`_

.. raw:: html

    <h1 align="center">
      <div style="position: relative; padding-bottom: 0%; overflow: hidden; max-width: 100%; height: auto;">
        <iframe width="600" height="480" src="https://www.youtube.com/embed/gKDsBsbIem4?autoplay=1" frameborder="1" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
      </div>
    </h1>

Overview
========

This tutorial shows how to simply utilize Speed Filter which is designed to limit the maximum speed of robots in speed restriction areas marked on a map. This functionality is being covered by ``SpeedFilter`` costmap filter plugin which will be enabled and used in this document.

Requirements
============

It is assumed that ROS 2, Gazebo and TurtleBot3 packages are installed or built locally. Please make sure that the Navigation2 project is also built locally as it was made in :ref:`build-instructions`.

Tutorial Steps
==============

1. Prepare filter mask
----------------------

> As was written in :ref:`concepts`, any Costmap Filter (including Speed Filter) are reading the data marked in a filter mask file. All information about filter masks, their types, detailed structure and how to make a new one is written in a :ref:`navigation2_with_keepout_filter` tutorial at "1. Prepare filter maks" chapter. The principal of drawing filter mask for Speed Filter is the same as for Keepout Filter, except that ``OccupancyGrid`` mask values have another meaning: these values are encoded speed limits for the areas corresponding to the cell on map.

Reviewer: ```suggestion
As was written in :ref:`concepts`, any Costmap Filter (including Speed Filter) is reading the data marked in a filter mask file. All information about filter masks, their types, detailed structure and how to make a new one is written in a :ref:`navigation2_with_keepout_filter` tutorial at ``1. Prepare filter masks`` chapter. The principal of drawing the filter mask for Speed Filter is the same as for Keepout Filter (to annotate a map with the requested zones), except that ``OccupancyGrid`` mask values have another meaning: these values are encoded speed limits for the areas corresponding to the cell on map.
```

Let's look, how it is being decoded. As we know, ``OccupancyGrid`` values are belonging to the ``[0..100]`` range. For Speed Filter ``0`` value means no speed limit in the area corresponding zero-cell on mask. Values from ``[1..100]`` range are being linearly converted into a speed limit value by the following formula:

.. code-block:: c++

  speed_limit = filter_mask_data * multiplier + base;

where: 

 - ``filter_mask_data`` - is an ``OccupancyGrid`` value of the corresponding cell on mask where maximum speed should be restricted.
 - ``base`` and ``multiplier`` are coefficients taken from ``nav2_msgs/CostmapFilterInfo`` messages published by Costmap Filter Info Server (see in next chapter below).

The decoded ``speed_limit`` value may have one of two meanings:

 - Speed limit expressed in a percent from maximum robot speed.
 - Speed limit expressed in absolute values (e.g. in ``m/s``).

The meaning used by Speed Filter is being read from ``nav2_msgs/CostmapFilterInfo`` messages.
In this tutorial we will use the first type of speed restriction expressed in a percent from maximum robot speed.

.. note::

>   For speed restriction expressed in a percent, ``speed_limit`` will be used exactly as a percent belonging to ``[0..100]`` range, but not as a share from ``[0.0..1.0]`` range.

Reviewer: ```suggestion
  For speed restriction expressed in a percent, ``speed_limit`` will be used exactly as a percent belonging to ``[0..100]`` range, not ``[0.0..1.0]`` range.
```

Create a new image with a PGM/PNG/BMP format: copy ``turtlebot3_world.pgm`` main map which will be used in a world simulation from a navigation2 repository to a new ``speed_mask.pgm`` file. Open ``speed_mask.pgm`` in your favourite raster graphics editor and fill speed restricted areas with grey colors. In our example darker colors will indicate areas with higher speed restriction:

.. image:: images/Navigation2_with_Speed_Filter/drawing_speed_mask.png
    :width: 500px

> Area "A" is filled with ``40%`` gray color, area "B" - with ``70%`` gray, what should mean that speed restriction will take ``100% - 40% = 60%`` in area "A" and ``100% - 70% = 30%`` in area "B" from maximum speed value allowed for this robot.

Reviewer: ```suggestion
Area "A" is filled with ``40%`` gray color, area "B" - with ``70%`` gray, that means that speed restriction will take ``100% - 40% = 60%`` in area "A" and ``100% - 70% = 30%`` in area "B" from maximum speed value allowed for this robot.
```

> We will use ``scale`` map mode with no thresholds. In this mode darker colors will have higher ``OccupancyGrid`` values. E.g. for area "B" with ``70%`` of gray ``OccupancyGrid`` data will be equal to ``70``. So in order to hit the target (darker colors indicate areas with higher speed restriction), we need to choose ``base = 100.0`` and ``multiplier = -1.0``. This will reverse the scale ``OccupancyGrid`` values to a desired one.

Reviewer: ```suggestion
We will use ``scale`` map mode with no thresholds. In this mode darker colors will have higher ``OccupancyGrid`` values. E.g. for area "B" with ``70%`` of gray ``OccupancyGrid`` data will be equal to ``70``. So in order to hit the target, we need to choose ``base = 100.0`` and ``multiplier = -1.0``. This will reverse the scale ``OccupancyGrid`` values to a desired one.
```

You already mentioned in the sentence prior

  Another important thing is that it is not necessary to use the whole ``[0..100]`` percent scale. ``base`` and ``multiplier`` coefficients could be chosen so that the speed restriction values would belong to somewhere in the middle of percent range. E.g. ``base = 40.0``, ``multiplier = 0.1`` will give speed restrictions from ``[40.0%..50.0%]`` range with a step of ``0.1%``. This might be useful for fine tuning.

After all speed restriction areas will be filled, save the ``speed_mask.pgm`` image.

Like all other maps, the filter mask should have its own YAML metadata file. Copy `turtlebot3_world.yaml <https://github.com/ros-planning/navigation2/blob/main/nav2_bringup/bringup/maps/turtlebot3_world.yaml>`_ to ``speed_mask.yaml``. Open ``speed_mask.yaml`` and update the fields as shown below (as mentioned before for the ``scale`` mode to use whole color lightness range there should be no thresholds: ``free_thresh = 0.0`` and ``occupied_thresh = 1.0``):

.. code-block:: yaml

  image: turtlebot3_world.pgm
  ->
  image: speed_mask.pgm

  mode: trinary
  ->
  mode: scale

  occupied_thresh: 0.65
  free_thresh: 0.196
  ->

>   occupied_thresh: 1.0

Reviewer: are these 2 thresholds even used in this situation? Is it important to change at all?

Author: Yes, that is because of in the example we want to have 1:1 mask grayscale % -> speed limit % conversion. From the tutorial:
> We will use scale map mode with no thresholds.

This mean, that `occupied_thresh = 1.0` and `free_thresh = 0.0`.

Reviewer: That should be made really clear why you're doing this and that its important, I would have just scanned right through that without thinking that it didn't apply to this as much.

Author: Added a sentence about thresholds choice in https://github.com/ros-planning/navigation.ros.org/pull/124/commits/26a9ceea8da06b9bdd8e5a00694e35ce46add883

  free_thresh: 0.0

Save ``speed_mask.yaml`` and the new filter mask is ready to use.

.. note::

>   World map itself and filter mask could have different sizes, origin and resolution which might be useful e.g. for cases when filter mask is covering smaller areas on maps or when one filter mask is used repeatedly many times (like annotating a speed restricted area for same shape rooms in the hotel). For this case, you need to correct ``resolution`` and ``origin`` fields in YAML as well so that the filter mask is correctly laid on top of the original map.

Reviewer: ```suggestion
  World map itself and filter mask could have different sizes, origin and resolution which might be useful (e.g. for cases when filter mask is covering smaller areas on maps or when one filter mask is used repeatedly many times, like annotating a speed restricted area for same shape rooms in the hotel). For this case, you need to correct ``resolution`` and ``origin`` fields in YAML as well so that the filter mask is correctly laid on top of the original map. This example shows using the main map as a base, but that is not required.
```

.. note::

>   Another important note is that since Costmap2D does not support orientation, the last third "yaw" component of the ``origin`` vector should be equal to zero. For example: ``origin: [1.25, -5.18, 0.0]``.

Reviewer: Remove this as a note and add it as a sentence in the yaml file modification section above

Author: Moved in https://github.com/ros-planning/navigation.ros.org/pull/124/commits/0ca276abe001d1fddab1d0caf89e8a9d15a95f5c#diff-8e354ab7d89ece0f0ce8893b7a0ac5fd4a8c11335c98fd9073499bb896162d95R93

              parameters=[configured_params])

      ld = LaunchDescription()

      ld.add_action(declare_namespace_cmd)
      ld.add_action(declare_use_sim_time_cmd)
      ld.add_action(declare_autostart_cmd)
      ld.add_action(declare_params_file_cmd)
      ld.add_action(declare_mask_yaml_file_cmd)

      ld.add_action(start_lifecycle_manager_cmd)
      ld.add_action(start_map_server_cmd)
      ld.add_action(start_costmap_filter_info_server_cmd)

      return ld

where the ``params_file`` variable should be set to a YAML-file having ROS parameters for Costmap Filter Info Publisher Server and Map Server nodes. These parameters and their meaning are listed at :ref:`configuring_map_server` page. Please, refer to it for more information. The example of ``params_file`` could be found below:

.. code-block:: yaml

  costmap_filter_info_server:
    ros__parameters:
      use_sim_time: true
      type: 1
      filter_info_topic: "/costmap_filter_info"
      mask_topic: "/filter_mask"
      base: 100.0
      multiplier: -1.0
  filter_mask_server:
    ros__parameters:
      use_sim_time: true
      frame_id: "map"

>       topic_name: "/filter_mask"

Reviewer: Lets chose a less generic name than this for the topic, keep in mind this is the tutorial that people will use as a design reference so lets use a topic that's actually indicative of speed filter

Author: Updated to `/speed_filter_mask` in https://github.com/ros-planning/navigation.ros.org/commit/0ca276abe001d1fddab1d0caf89e8a9d15a95f5c#diff-8e354ab7d89ece0f0ce8893b7a0ac5fd4a8c11335c98fd9073499bb896162d95L226.
Plus updated in the same way for the Keepout tutorial: https://github.com/ros-planning/navigation.ros.org/commit/0ca276abe001d1fddab1d0caf89e8a9d15a95f5c#diff-9de6c0db49184c10da140edc0e8a0b610ce2186d7c72119697d77b581184d93cL208.

Also, updated the tutorials repo accordingly to fit the documentation:
https://github.com/ros-planning/navigation2_tutorials/pull/18/commits/284e4bed38804d582b479dc4b4a1fc008f519921

One thing was untouched - is a demo videos for Keepout and Speed filter. I do not think that to re-take whole video only because of parameter name change is a reasonable idea. However, correct me if I'm wrong.

Reviewer: yeah, don't update a video for small details.

      yaml_filename: "speed_mask.yaml"

Note, that:

 - For Speed Filter setting speed restrictions in a percent from maximum speed, the ``type`` of costmap filter should be set to ``1``. All possible costmap filter types could be found at :ref:`configuring_map_server` page.
 - Filter mask topic name should be the equal for ``mask_topic`` parameter of Costmap Filter Info Publisher Server and ``topic_name`` parameter of Map Server.

>  - As was described in a previous chapter, ``base`` and ``multiplier`` should be set to ``100.0`` and ``-1.0`` accordingly.

Reviewer: ```suggestion
 - As was described in a previous chapter, ``base`` and ``multiplier`` should be set to ``100.0`` and ``-1.0`` accordingly for the purposes of this tutorial example.
```

  $ cd ~/tutorials_ws/src
  $ git clone https://github.com/ros-planning/navigation2_tutorials.git
  $ cd ~/tutorials_ws
  $ colcon build --symlink-install --packages-select nav2_costmap_filters_demo
  $ source ~/tutorials_ws/install/setup.bash
  $ ros2 launch nav2_costmap_filters_demo costmap_filter_info.launch.py params_file:=src/navigation2_tutorials/nav2_costmap_filters_demo/params/speed_params.yaml mask:=src/navigation2_tutorials/nav2_costmap_filters_demo/maps/speed_mask.yaml

3. Enable Speed Filter
----------------------

Costmap Filters are Costamp2D plugins. You can enable the ``SpeedFilter`` plugin in Costmap2D by adding ``speed_filter`` to the ``plugins`` parameter in ``nav2_params.yaml``. The Speed Filter plugin should have the following parameters defined:

- ``plugin``: type of plugin. In our case ``nav2_costmap_2d::SpeedFilter``.
- ``filter_info_topic``: filter info topic name. This needs to be equal to ``filter_info_topic`` parameter of Costmap Filter Info Publisher Server from the chapter above.
- ``speed_limit_topic``: name of topic to publish speed limit to.

Full list of parameters supported by ``SpeedFilter`` are listed at the :ref:`speed_filter` page.

> You can place the plugin either in the ``global_costmap`` section in ``nav2_params.yaml`` to have speed restriction mask applied to global costmap or in the ``local_costmap`` to apply speed mask to the local costmap. However, it is highly not recommended to enable ``SpeedFilter`` plugin simultaneously for global and local costmaps. This might produce unwanted multiple "speed restriction" - "no restriction" message chains on speed restriction boundaries, that will cause jerking of the robot or another unpredictable behaviour.

Reviewer: Please be more stern than a recommendation, we do not want anyone to have them in both because it would really mess things up

Author: This was changed to ban starting from https://github.com/ros-planning/navigation.ros.org/pull/124/commits/0ca276abe001d1fddab1d0caf89e8a9d15a95f5c#diff-8e354ab7d89ece0f0ce8893b7a0ac5fd4a8c11335c98fd9073499bb896162d95L265

> As stated in the `design <https://github.com/ros-planning/navigation2/blob/main/doc/design/CostmapFilters_design.pdf>`_, Speed Filter publishes speed restricting `messages <https://github.com/ros-planning/navigation2/blob/main/nav2_msgs/msg/SpeedLimit.msg>`_ targeted for a Controller Server so that it could restrict maximum speed of the robot when it needed. Controller Server has a ``speed_limit_topic`` ROS parameter for that, which should be set to the same as in ``speed_filter`` plugin value.

Reviewer: ```suggestion
As stated in the `design <https://github.com/ros-planning/navigation2/blob/main/doc/design/CostmapFilters_design.pdf>`_, Speed Filter publishes speed restricting `messages <https://github.com/ros-planning/navigation2/blob/main/nav2_msgs/msg/SpeedLimit.msg>`_ targeted for a Controller Server so that it could restrict maximum speed of the robot when it needed. Controller Server has a ``speed_limit_topic`` ROS parameter for that, which should be set to the same as in ``speed_filter`` plugin value. This topic in the map server could also be used to any number of other speed-restricted applications beyond the speed limiting zones, such as dynamically adjusting maximum speed by payload mass.
```

To enable Speed Filter with same mask for both global costmap, use the following configuration:

.. code-block:: text

  global_costmap:
    global_costmap:
      ros__parameters:
        ...
        plugins: ["static_layer", "obstacle_layer", "inflation_layer", "speed_filter"]
        ...
        speed_filter:
          plugin: "nav2_costmap_2d::SpeedFilter"
          enabled: True
          filter_info_topic: "/costmap_filter_info"
          speed_limit_topic: "/speed_limit"

> For using Speed Filter with local costmap, use the following configuration:

Reviewer: We should choose one or the other local / global costmap to recommend in the tutorial.

Author: Fixed stating from https://github.com/ros-planning/navigation.ros.org/commit/0ca276abe001d1fddab1d0caf89e8a9d15a95f5c#diff-8e354ab7d89ece0f0ce8893b7a0ac5fd4a8c11335c98fd9073499bb896162d95L269

Also, arranged some sentences there to make the narration more consistent:
https://github.com/ros-planning/navigation.ros.org/pull/124/commits/0ca276abe001d1fddab1d0caf89e8a9d15a95f5c#diff-8e354ab7d89ece0f0ce8893b7a0ac5fd4a8c11335c98fd9073499bb896162d95L267
->
https://github.com/ros-planning/navigation.ros.org/pull/124/commits/0ca276abe001d1fddab1d0caf89e8a9d15a95f5c#diff-8e354ab7d89ece0f0ce8893b7a0ac5fd4a8c11335c98fd9073499bb896162d95R279
